// @vitejs/plugin-react-oxc/dist/index.mjs
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { readFileSync } from 'node:fs';

const runtimePublicPath = "/@react-refresh";
const reactCompRE = /extends\s+(?:React\.)?(?:Pure)?Component/;
const refreshContentRE = /\$Refresh(?:Reg|Sig)\$\(/;
const preambleCode = `import { injectIntoGlobalHook } from "__BASE__${runtimePublicPath.slice(
  1
)}"
injectIntoGlobalHook(window);
window.$RefreshReg$ = () => {};
window.$RefreshSig$ = () => (type) => type;`;
const getPreambleCode = (base) => preambleCode.replace("__BASE__", base);
const avoidSourceMapOption = Symbol();
function addRefreshWrapper(code, map, pluginName, id) {
  const hasRefresh = refreshContentRE.test(code);
  const onlyReactComp = !hasRefresh && reactCompRE.test(code);
  const normalizedMap = map === avoidSourceMapOption ? null : map;
  if (!hasRefresh && !onlyReactComp) return { code, map: normalizedMap };
  const avoidSourceMap = map === avoidSourceMapOption;
  const newMap = typeof normalizedMap === "string" ? JSON.parse(normalizedMap) : normalizedMap;
  let newCode = code;
  if (hasRefresh) {
    const refreshHead = removeLineBreaksIfNeeded(
      `let prevRefreshReg;
let prevRefreshSig;

if (import.meta.hot && !inWebWorker) {
  if (!window.$RefreshReg$) {
    throw new Error(
      "${pluginName} can't detect preamble. Something is wrong."
    );
  }

  prevRefreshReg = window.$RefreshReg$;
  prevRefreshSig = window.$RefreshSig$;
  window.$RefreshReg$ = RefreshRuntime.getRefreshReg(${JSON.stringify(id)});
  window.$RefreshSig$ = RefreshRuntime.createSignatureFunctionForTransform;
}

`,
      avoidSourceMap
    );
    newCode = `${refreshHead}${newCode}

if (import.meta.hot && !inWebWorker) {
  window.$RefreshReg$ = prevRefreshReg;
  window.$RefreshSig$ = prevRefreshSig;
}
`;
    if (newMap) {
      newMap.mappings = ";".repeat(16) + newMap.mappings;
    }
  }
  const sharedHead = removeLineBreaksIfNeeded(
    `import * as RefreshRuntime from "${runtimePublicPath}";
const inWebWorker = typeof WorkerGlobalScope !== 'undefined' && self instanceof WorkerGlobalScope;

`,
    avoidSourceMap
  );
  newCode = `${sharedHead}${newCode}

if (import.meta.hot && !inWebWorker) {
  RefreshRuntime.__hmr_import(import.meta.url).then((currentExports) => {
    RefreshRuntime.registerExportsForReactRefresh(${JSON.stringify(
    id
  )}, currentExports);
    import.meta.hot.accept((nextExports) => {
      if (!nextExports) return;
      const invalidateMessage = RefreshRuntime.validateRefreshBoundaryAndEnqueueUpdate(${JSON.stringify(
    id
  )}, currentExports, nextExports);
      if (invalidateMessage) import.meta.hot.invalidate(invalidateMessage);
    });
  });
}
`;
  if (newMap) {
    newMap.mappings = ";;;" + newMap.mappings;
  }
  return { code: newCode, map: newMap };
}
function removeLineBreaksIfNeeded(code, enabled) {
  return enabled ? code.replace(/\n/g, "") : code;
}

const silenceUseClientWarning = (userConfig) => ({
  rollupOptions: {
    onwarn(warning, defaultHandler) {
      if (warning.code === "MODULE_LEVEL_DIRECTIVE" && warning.message.includes("use client")) {
        return;
      }
      if (warning.code === "SOURCEMAP_ERROR" && warning.message.includes("resolve original location") && warning.pos === 0) {
        return;
      }
      if (userConfig.build?.rollupOptions?.onwarn) {
        userConfig.build.rollupOptions.onwarn(warning, defaultHandler);
      } else {
        defaultHandler(warning);
      }
    }
  }
});

const _dirname = dirname(fileURLToPath(import.meta.url));
const refreshRuntimePath = join(_dirname, "refresh-runtime.js") ;
const defaultIncludeRE = /\.[tj]sx?(?:$|\?)/;
function viteReact(opts = {}) {
  const include = opts.include ?? defaultIncludeRE;
  const exclude = [
    ...Array.isArray(opts.exclude) ? opts.exclude : opts.exclude ? [opts.exclude] : [],
    /\/node_modules\//
  ];
  const jsxImportSource = opts.jsxImportSource ?? "react";
  const jsxImportRuntime = `${jsxImportSource}/jsx-runtime`;
  const jsxImportDevRuntime = `${jsxImportSource}/jsx-dev-runtime`;
  const viteConfig = {
    name: "vite:react-oxc:config",
    config(userConfig, { command }) {
      return {
        // @ts-expect-error rolldown-vite Vite type incompatibility
        build: silenceUseClientWarning(userConfig),
        oxc: {
          jsx: {
            runtime: "automatic",
            importSource: jsxImportSource,
            refresh: command === "serve",
            development: command === "serve"
          },
          jsxRefreshInclude: include,
          jsxRefreshExclude: exclude
        },
        optimizeDeps: {
          include: [
            "react",
            "react-dom",
            jsxImportDevRuntime,
            jsxImportRuntime
          ],
          rollupOptions: { jsx: { mode: "automatic" } }
        }
      };
    },
    options() {
      if (!this.meta.rolldownVersion) {
        throw new Error(
          "@vitejs/plugin-react-oxc requires rolldown-vite to be used. See https://vitejs.dev/guide/rolldown for more details about rolldown-vite."
        );
      }
    }
  };
  const viteRefreshRuntime = {
    name: "vite:react-oxc:refresh-runtime",
    enforce: "pre",
    resolveId: {
      filter: { id: exactRegex(runtimePublicPath) },
      handler(id) {
        return id;
      }
    },
    load: {
      filter: { id: exactRegex(runtimePublicPath) },
      handler(_id) {
        return readFileSync(refreshRuntimePath, "utf-8").replace(
          /__README_URL__/g,
          "https://github.com/vitejs/vite-plugin-react/tree/main/packages/plugin-react-oxc"
        );
      }
    }
  };
  let skipFastRefresh = false;
  const viteRefreshWrapper = {
    name: "vite:react-oxc:refresh-wrapper",
    apply: "serve",
    configResolved(config) {
      skipFastRefresh = config.isProduction || config.server.hmr === false;
    },
    transform: {
      filter: {
        id: { include, exclude }
      },
      handler(code, id, options) {
        const ssr = options?.ssr === true;
        const [filepath] = id.split("?");
        const isJSX = filepath.endsWith("x");
        const useFastRefresh = !skipFastRefresh && !ssr && (isJSX || code.includes(jsxImportDevRuntime) || code.includes(jsxImportRuntime));
        if (!useFastRefresh) return;
        const { code: newCode } = addRefreshWrapper(
          code,
          avoidSourceMapOption,
          "@vitejs/plugin-react-oxc",
          id
        );
        return { code: newCode, map: null };
      }
    },
    transformIndexHtml(_, config) {
      if (!skipFastRefresh)
        return [
          {
            tag: "script",
            attrs: { type: "module" },
            children: getPreambleCode(config.server.config.base)
          }
        ];
    }
  };
  return [viteConfig, viteRefreshRuntime, viteRefreshWrapper];
}
function exactRegex(input) {
  return new RegExp(`^${escapeRegex(input)}$`);
}
const escapeRegexRE = /[-/\\^$*+?.()|[\]{}]/g;
function escapeRegex(str) {
  return str.replace(escapeRegexRE, "\\$&");
}

export { viteReact as default };
